File: autoconduct/resumer.py
"""Resume a stalled session headlessly via the claude CLI."""

from __future__ import annotations

import subprocess
from dataclasses import dataclass

from .conductor_db import StalledSession
from .config import Config

RESUME_TIMEOUT_SECONDS = 60 * 60  # one long agentic run, but never forever


@dataclass(frozen=True)
class ResumeResult:
    session_id: str
    ok: bool
    detail: str


def resume_session(
    session: StalledSession, config: Config, dry_run: bool = False
) -> ResumeResult:
    """Run `claude --resume` in the session's workspace directory.

    Headless print mode (-p) so the run terminates on its own. Work lands
    in the workspace files/git; Conductor's chat UI won't show the turns.
    """
    cmd = [
        "claude",
        "--resume",
        session.claude_session_id,
        "-p",
        config.resume_prompt,
        "--permission-mode",
        config.permission_mode,
    ]
    if dry_run:
        return ResumeResult(session.session_id, True, f"DRY RUN: {' '.join(cmd)}")
    try:
        proc = subprocess.run(
            cmd,
            cwd=session.workspace_path,
            capture_output=True,
            text=True,
            timeout=RESUME_TIMEOUT_SECONDS,
        )
    except subprocess.TimeoutExpired:
        return ResumeResult(session.session_id, False, "timed out after 1h")
    except FileNotFoundError:
        return ResumeResult(session.session_id, False, "claude CLI not found on PATH")
    if proc.returncode != 0:
        tail = (proc.stderr or proc.stdout or "").strip()[-300:]
        return ResumeResult(session.session_id, False, f"exit {proc.returncode}: {tail}")
    return ResumeResult(session.session_id, True, (proc.stdout or "").strip()[-300:])

File: NightConductor/Sources/NightConductor/Resumer.swift
import Foundation

struct ResumeResult {
    let ok: Bool
    let detail: String
}

/// Thread-safe accumulator for a pipe drained on a background queue.
private final class OutputBox: @unchecked Sendable {
    private let lock = NSLock()
    private var data = Data()
    func append(_ chunk: Data) { lock.lock(); data.append(chunk); lock.unlock() }
    var string: String {
        lock.lock(); defer { lock.unlock() }
        return String(data: data, encoding: .utf8) ?? ""
    }
}

/// Resumes a stalled session headlessly via `claude --resume` in the
/// session's workspace directory.
enum Resumer {
    static let timeoutSeconds: TimeInterval = 3600 // one long agentic run, never forever

    /// GUI apps get a minimal PATH, so discover the claude binary the way a
    /// terminal would: common install locations first, then a login shell
    /// (which picks up nvm/asdf setups).
    static func findClaudeBinary() -> String? {
        let candidates = [
            NSHomeDirectory() + "/.claude/local/claude",
            "/opt/homebrew/bin/claude",
            "/usr/local/bin/claude",
            NSHomeDirectory() + "/.local/bin/claude",
        ]
        for path in candidates where FileManager.default.isExecutableFile(atPath: path) {
            return path
        }
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/zsh")
        process.arguments = ["-lc", "command -v claude"]
        let stdout = Pipe()
        process.standardOutput = stdout
        process.standardError = Pipe()
        guard (try? process.run()) != nil else { return nil }
        process.waitUntilExit()
        guard process.terminationStatus == 0 else { return nil }
        let output = String(
            data: stdout.fileHandleForReading.readDataToEndOfFile(),
            encoding: .utf8
        )?.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let path = output, FileManager.default.isExecutableFile(atPath: path) else {
            return nil
        }
        return path
    }

    static func resume(
        session: StalledSession,
        claudePath: String,
        config: PolicyConfig
    ) -> ResumeResult {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: claudePath)
        process.arguments = [
            "--resume", session.claudeSessionID,
            "-p", config.resumePrompt,
            "--permission-mode", config.permissionMode,
        ]
        process.currentDirectoryURL = URL(fileURLWithPath: session.workspacePath)

        // claude is typically a node script; make sure its interpreter is
        // reachable by prepending the binary's own directory to PATH.
        var environment = ProcessInfo.processInfo.environment
        let claudeDir = (claudePath as NSString).deletingLastPathComponent
        let basePath = environment["PATH"] ?? "/usr/bin:/bin"
        environment["PATH"] = "\(claudeDir):/opt/homebrew/bin:/usr/local/bin:\(basePath)"
        process.environment = environment

        let stdout = Pipe()
        let stderr = Pipe()
        process.standardOutput = stdout
        process.standardError = stderr

        let finished = DispatchSemaphore(value: 0)
        process.terminationHandler = { _ in finished.signal() }

        do {
            try process.run()
        } catch {
            return ResumeResult(ok: false, detail: "cannot launch claude: \(error.localizedDescription)")
        }

        // Drain BOTH pipes concurrently while the process runs. A long
        // agentic run emits enough stderr to fill the 64KB pipe buffer; if
        // we only read after exit, the child blocks writing and never exits.
        let outBox = OutputBox()
        let errBox = OutputBox()
        let readers = DispatchGroup()
        let queue = DispatchQueue(label: "nightconductor.resume.read", attributes: .concurrent)
        for (pipe, box) in [(stdout, outBox), (stderr, errBox)] {
            readers.enter()
            queue.async {
                box.append(pipe.fileHandleForReading.readDataToEndOfFile())
                readers.leave()
            }
        }

        if finished.wait(timeout: .now() + timeoutSeconds) == .timedOut {
            process.terminate()
            readers.wait() // pipes hit EOF once the child is gone
            return ResumeResult(ok: false, detail: "timed out after 1h")
        }
        readers.wait()

        let combined = (outBox.string + errBox.string).trimmingCharacters(in: .whitespacesAndNewlines)
        if process.terminationStatus != 0 {
            return ResumeResult(ok: false, detail: "exit \(process.terminationStatus): \(combined.suffix(200))")
        }
        return ResumeResult(ok: true, detail: String(outBox.string.trimmingCharacters(in: .whitespacesAndNewlines).suffix(200)))
    }
}
